docs/book/src/developer/providers/contracts/infra-machinepool.md

```diff
@@ -0,0 +1,542 @@
+
+In case necessary, "terminal failures" should be surfaced using conditions, with a well documented type/reason;
+it is up to consumers to treat them accordingly.
+
+</aside>
+
+### InfraMachinePoolTemplate, InfraMachineTemplatePoolList resource definition
+
+For a given InfraMachinePool resource, you SHOULD also add a corresponding InfraMachinePoolTemplate resource in order to use it in ClusterClasses. The template resource MUST be name `<InfraMachinePool>Template`.
+
+
+```go
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=foomachinepooltemplates,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// FooMachinePoolTemplate is the Schema for the foomachinepooltemplates API.
+type FooMachinePoolTemplate struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec FooMachinePoolTemplateSpec `json:"spec,omitempty"`
+}
+
+type FooMachinePoolTemplateSpec struct {
+    Template FooMachinePooleTemplateResource `json:"template"`
+}
+
+type FooMachinePoolTemplateResource struct {
+    // Standard object's metadata.
+    // More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+    // +optional
+    ObjectMeta clusterv1.ObjectMeta `json:"metadata,omitempty,omitzero"`
+    Spec FooMachinePoolSpec `json:"spec"`
+}
+```
+
+NOTE: in this example `spec.template.spec` embeds `FooMachinePoolSpec` from MachinePool. This might not always be
+the best choice depending of if/how InfraMachinePools spec fields applies to many machine pools vs only one.
+
+For each InfraMachinePoolTemplate resource, you MUST also add the corresponding list resource.
+The list resource MUST be named as `<InfraMachinePoolTemplate>List`.
+
+```go
+// +kubebuilder:object:root=true
+
+// FooMachinePoolTemplateList contains a list of FooMachinePoolTemplates.
+type FooMachinePoolTemplateList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []FooMachinePoolTemplate `json:"items"`
+}
+```
+
+### InfraMachinePoolTemplate: support for SSA dry run
+
+When Cluster API's topology controller is trying to identify differences between templates defined in a ClusterClass and
+the current Cluster topology, it is required to run [Server Side Apply] (SSA) dry run call.
+
+However, in case you have immutability checks for your InfraMachinePoolTemplate, this can lead the SSA dry run call to error.
+
+In order to avoid this InfraMachinePoolTemplate MUST specifically implement support for SSA dry run calls from the topology controller.
+
+The implementation requires to use controller runtime's `CustomValidator`, available in CR versions >= v0.12.3.
```

Comment:
```suggestion
The implementation requires to use controller-runtime's `CustomValidator`, available since version v0.12.3.
```

(CR often stands for "Custom Resource", not for controller-runtime)

Reply:
Goo dpoint, changed.